jsonschema-core.xml

```diff
@@ -608,6 +608,38 @@
                 </t>
             </section>
 
+            <section title="Evaluation Results">
+                <t>
+                    Evaluating an instance against a schema MUST produce one of four general outcomes:
+
+                    <list>
+                        <t>
+                            success: the evaluation completed without failing any validation assertions
+                        </t>
+                        <t>
+                            failure: the evaluation completed with assertion failures
+                        </t>
+                        <t>
+                            runtime error: the schema was understood but the evaluation could not
+                            complete
+                        </t>
+                        <t>
+                            refusal to process: the schema was not understood and evaluation never began
+                        </t>
+                    </list>
+                </t>
+                <t>
+                    Runtime errors include but are not limited to:  failure of reference resolution,
+                    detection of an infinite reference loop, or excessive consumption of memory.
+                </t>
+                <t>
+                    Unlike runtime errors, refusal-to-process is the expected outcome in certain cases.
```

Comment:
I stil think this line is a bit fuzzy.  Consider parsing a number from a string.  Failing to parse `"123A5"` could be considered a refusal to process because

- the input was invalid, or
- the request to parse should have specified hexadecimal

In both cases, the outcome is failure external to the normal processing of a function.

Similarly, refusal to process and runtime failures both have an outcome of a failure that is external to the normal processing of a schema.

Reply:
@gregsdennis what do you suggest?  There is clearly a difference between "I asked for a feature to be supported and it's not so I'm going to do the sensible thing and not use this schema" and "oops something that was expected to work didn't work."  What way of describing these things would you find appropriate?

Reply:
That's my point: the difference is semantic.  It's the difference between two errors, and I don't think that the specification needs to define that difference.  The specification just needs to say, "validation cannot proceed in this case," and let the implementation decide what that means for its language.

A feature not being supported is not expected.  (In fact, .Net has `NotSupportedException` for that specific case.)  I can't think of reason why you'd want an application where evaluation is required to continue processing if that evaluation errors, whether it's for bad JSON, an unresolvable `$ref`, or an unknown vocab.  It's all the same outcome: something is wrong and the application has to stop (or gracefully catch the error and inform the user).

Reply:
There are two phases here:  "Can I use this?" and "I'm using this."  The way to react to failures in each is different.

What is the advantage in preventing these from being distinct?

Reply:
The other part of this is that the spec intentionally requires a ["refuse to process" behavior](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#section-8.1.2).

What's up for debate is not whether that is a behavior, as that was part of a past debate and was written into the spec.  What is up for debate now is how to explain "refuse to process" compared to other conditions.

PRs are not the place to re-litigate past decisions.  This PR just needs to clarify what's there.  Changing what's there requires a new issue.

Reply:
> The way to react to failures in each is different.

What are the different outcomes?  Both stop the implementation unexpectedly.

> What is the advantage in preventing these from being distinct?

What's the advantage of having them distinct?

> What's up for debate is not whether that is a behavior

I'm not saying it's not a behavior.  I'm saying the _outcome_ of that behavior is no different than any other failure to process error.  If the outcome is indistinguishable, then why distinguish the behavior?

Reply:
They are different outcomes.  One is directed at _the person who tried to use the schema_ and tells them to go load a plugin or something.  It is specifically called out in the spec as a distinct outcome.  The other types of errors tell _the person who wrote the schema_ that something is wrong with the schema.

* Different treatment in the specification language
* Different audiences for the errors
* Different types of corrective action
* Different characterization of the schema (incorrectly written vs correctly written)

These are different things.  In every dimension I can think of.  BTW, I do not care what form the errors take.  You can just throw different exceptions, that's fine with me- nothing in this PR has anything to do with _how_ they are reported.  But everything about these categories of outcomes is different.

Reply:
> One is directed at the person who tried to use the schema and tells them to go load a plugin or something. It is specifically called out in the spec as a distinct outcome. The other types of errors tell the person who wrote the schema that something is wrong with the schema.

When evaluating a schema, the author of the schema is (theoretically) uninvolved and likely will not receive notification that something went wrong.  The implementation's only recourse is to halt and report to the user why it halted.

Maybe it halts _differently_, I guess, but it still halts.  Whether the implementation _chooses_ to not process a schema or it _can't_ process a schema is irrelevant.  The outcome is that the schema isn't processed and must report some sort of failure.  The messaging in the failure tells the user

- the nature of the problem ("Different characterization of the schema (incorrectly written vs correctly written)")
- if they can do anything about it ("Different audiences for the errors")
- what to do ("Different types of corrective action")

It's just like a JSON parser that _can_ handle comments but isn't configured to do so.  When it encounters comments, it stops and reports what's wrong.

My point is that processing halts.  I don't think that the specification needs to say anything more than that.

Are you saying that the specification should prescribe the (minimum) content of a failure (if not the mechanism, e.g. an exception)?

Reply:
Or.... are you saying that you want to distinguish between errors that must be caught at evaluation-time and errors which _could_ be caught in some sort of static analysis?  Would that require implementations to support static analysis (mine doesn't)?

Reply:
That's how I view it -- there are two main kinds of exceptions:
- compile time exceptions: something is written incorrectly in the schema (mostly things we would catch by evaluating against the metaschema, although there are a few things in the spec that can't quite be expressed with the metaschema, as we've discovered.. no I'm not looking up a reference to insert here). These things can all be determined by statically inspecting the schema in isolation from any data instance, and, I think it could be argued, also in isolation from any particular implementation
- runtime exceptions: something went wrong while actually doing something with the schema (e.g. evaluating it against a data instance, using it to generate code or documentation), etc. some of these things might be localized to a particular implementation at a point in time (e.g. you forgot to load a supporting library that would let you validate domain names), or a particular state of the local resources (you have no internet connection, so you can't resolve a $ref that points to a non-local resource).

It might also be worthwhile to distinguish between transient and permanent errors. In my implementation, I try to only error at "compile time" for things that are not correctable (e.g. syntax errors), and anything that _could_ be corrected between schema-loading time and data-instance-evaluation time is deferred until runtime (loading optional modules for format handling, checking if the URI in a $ref actually resolves to a known schema location, etc).

> Would that require implementations to support static analysis (mine doesn't)?

When you load a schema into your implementation, do you generate an error if the schema is malformed (e.g. `"allOf":"not an array"`? Or do you do nothing with it until it's time to evaluate against a data instance?
You can also inspect a schema for errors simply by evaluating it against its metaschema.

Reply:
> When you load a schema into your implementation, do you generate an error if the schema is malformed (e.g. "allOf":"not an array"?

Yes, but they would be considered a deserialization (JSON text -> model) error.  I don't see that as static analysis.

I think for this discussion we can assume that we have a syntactically valid schema.

Evaluation against a meta-schema does (optionally) happen, but a lot of that is baked into the deserialization because the models are strongly typed and literally _can't_ accept invalid values.